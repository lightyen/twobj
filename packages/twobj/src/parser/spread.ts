import * as nodes from "./nodes"
import * as parser from "./parse_regexp"

export type SpreadDescription = {
	target: nodes.Classname | nodes.ArbitraryClassname | nodes.ArbitraryProperty | nodes.ShortCss
	value: string
	variants: nodes.Variant[]
	important: boolean
}

export function spread(text: string, { separator = ":" }: { separator?: string } = {}) {
	interface Context {
		variants: nodes.Variant[]
		important: boolean
	}

	const items: SpreadDescription[] = []
	const emptyGroup: nodes.Group[] = []
	const emptyVariants: nodes.VariantSpan[] = []
	const notClosed: nodes.BracketNode[] = []

	const walk = (node: nodes.TwExpression | undefined, ctx: Context): void => {
		if (node == undefined) return

		if (nodes.NodeType.VariantSpan === node.type) {
			if (node.child == undefined) {
				emptyVariants.push(node)
				return
			}
			const variants = ctx.variants.slice()
			variants.push(node.variant)
			walk(node.child, { ...ctx, variants })
		} else if (nodes.NodeType.Group === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			if (node.expressions.length === 0) {
				emptyGroup.push(node)
				return
			}

			node.expressions.forEach(n => walk(n, { ...ctx, important: ctx.important || node.important }))
		} else if (nodes.NodeType.ArbitraryClassname === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			if (node.e && node.e.type === nodes.NodeType.WithOpacity && node.e.closed === false) {
				notClosed.push(node.e)
				return
			}

			items.push({
				target: node,
				value: text.slice(...node.range),
				...ctx,
				important: ctx.important || node.important,
			})
		} else if (nodes.NodeType.ArbitraryProperty === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			items.push({
				target: node,
				value: text.slice(...node.range),
				...ctx,
				important: ctx.important || node.important,
			})
		} else if (nodes.NodeType.ShortCss === node.type) {
			if (!node.closed) {
				// notClosed.push(node)
				return
			}

			items.push({
				target: node,
				value: text.slice(...node.range),
				...ctx,
				important: ctx.important || node.important,
			})
		} else if (nodes.NodeType.ClassName === node.type) {
			items.push({
				target: node,
				value: node.value,
				...ctx,
				important: ctx.important || node.important,
			})
		}
	}

	parser.setSeparator(separator)
	const program = parser.parse(text)
	program.expressions.forEach(expr => walk(expr, { variants: [], important: false }))

	return { items, emptyGroup, emptyVariants, notClosed } as const
}
