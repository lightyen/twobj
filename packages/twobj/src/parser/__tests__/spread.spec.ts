import { NodeType } from ".."
import { spread } from "../spread"

it("parser util spread", () => {
	const result = spread(`bg-gray-500
		hover:(border-4 border-blue-500)
		[
			.group:hover,
			// .group:focus]
		]:!(text-red-700)// bg-primary
		/* text-error ()
		*/()
		custom/30 text-yellow-600/80 bg-gray-500/[.3]
		bg-[url(http://example)]
		text-[rgba(3 3 3 / .8)]/[.2] sm:hover: ( q-test`)
	expect(result.items.map(i => i.target.type)).toEqual([
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
	])

	expect(result).toMatchSnapshot()
})

it("spread all types", () => {
	const text = `
		class-value
		(class-value)
		hover:class-value
		hover:(class-value)
		[]:class-value
		hover:[]:class-value
		any-[]:class-value
		hover:any-[]:class-value
		class-[value]
		class-value/opacity
		class-[value]/[opacity]
		class-[value]/opacity
		class-value/[opacity]
		[prop: value]

		// important prefix
		!class-value
		!(class-value)
		hover:!class-value
		hover:!(class-value)
		[]:!class-value
		hover:[]:!class-value
		any-[]:!class-value
		hover:any-[]:!class-value
		!class-[value]
		!class-value/opacity
		!class-[value]/[opacity]
		!class-[value]/opacity
		!class-value/[opacity]
		![prop: value]

		// important after
		class-value!
		(class-value)!
		hover:class-value!
		hover:(class-value)!
		[]:class-value!
		hover:[]:class-value!
		any-[]:class-value!
		hover:any-[]:class-value!
		class-[value]!
		class-value/opacity!
		class-[value]/[opacity]!
		class-[value]/opacity!
		class-value/[opacity]!
		[prop: value]!
	`
	expect(spread(text).items.map(i => i.target.type)).toEqual([
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryProperty,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryProperty,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryProperty,
	])

	expect(spread(text).items.map(i => i.important)).toEqual([
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
	])

	expect(spread(text).items.map(i => i.target.important)).toEqual([
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		true,
		false,
		true,
		false,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		false,
		true,
		false,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
		true,
	])
})

it("parser arbitrary variant", () => {
	const result = spread(`tab-[]:black`)
	expect(result).toMatchSnapshot()
})
