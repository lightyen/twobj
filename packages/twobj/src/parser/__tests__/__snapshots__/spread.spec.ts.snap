// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser arbitrary variant 1`] = `
Object {
  "emptyGroup": Array [],
  "emptyVariants": Array [],
  "items": Array [
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          7,
          12,
        ],
        "type": "ClassName",
        "value": "black",
      },
      "value": "black",
      "variants": Array [
        Object {
          "range": Array [
            0,
            6,
          ],
          "selector": Object {
            "range": Array [
              5,
              5,
            ],
            "type": "CssSelector",
            "value": "",
          },
          "type": "ArbitraryVariant",
          "value": "tab-[]",
        },
      ],
    },
  ],
  "notClosed": Array [],
}
`;

exports[`parser util spread 1`] = `
Object {
  "emptyGroup": Array [
    Object {
      "closed": true,
      "expressions": Array [],
      "important": false,
      "range": Array [
        144,
        146,
      ],
      "type": "Group",
    },
  ],
  "emptyVariants": Array [
    Object {
      "range": Array [
        256,
        262,
      ],
      "type": "VariantSpan",
      "variant": Object {
        "id": Object {
          "range": Array [
            256,
            261,
          ],
          "type": "Identifier",
          "value": "hover",
        },
        "range": Array [
          256,
          262,
        ],
        "type": "SimpleVariant",
      },
    },
  ],
  "items": Array [
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          0,
          11,
        ],
        "type": "ClassName",
        "value": "bg-gray-500",
      },
      "value": "bg-gray-500",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          21,
          29,
        ],
        "type": "ClassName",
        "value": "border-4",
      },
      "value": "border-4",
      "variants": Array [
        Object {
          "range": Array [
            14,
            19,
          ],
          "type": "SimpleVariant",
          "value": "hover",
        },
      ],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          30,
          45,
        ],
        "type": "ClassName",
        "value": "border-blue-500",
      },
      "value": "border-blue-500",
      "variants": Array [
        Object {
          "range": Array [
            14,
            19,
          ],
          "type": "SimpleVariant",
          "value": "hover",
        },
      ],
    },
    Object {
      "important": true,
      "target": Object {
        "important": false,
        "range": Array [
          94,
          106,
        ],
        "type": "ClassName",
        "value": "text-red-700",
      },
      "value": "text-red-700",
      "variants": Array [
        Object {
          "range": Array [
            50,
            90,
          ],
          "selector": Object {
            "range": Array [
              50,
              90,
            ],
            "type": "CssSelector",
            "value": "
			.group:hover,
			// .group:focus]
		",
          },
          "type": "ArbitrarySelector",
          "value": "
			.group:hover,
			// .group:focus]
		",
        },
      ],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          149,
          158,
        ],
        "type": "ClassName",
        "value": "custom/30",
      },
      "value": "custom/30",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          159,
          177,
        ],
        "type": "ClassName",
        "value": "text-yellow-600/80",
      },
      "value": "text-yellow-600/80",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": Object {
          "closed": true,
          "opacity": Object {
            "range": Array [
              191,
              193,
            ],
            "type": "Identifier",
            "value": ".3",
          },
          "range": Array [
            190,
            194,
          ],
          "type": "WithOpacity",
        },
        "expr": undefined,
        "important": false,
        "prefix": Object {
          "range": Array [
            178,
            189,
          ],
          "type": "Identifier",
          "value": "bg-gray-500",
        },
        "range": Array [
          178,
          194,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "bg-gray-500/[.3]",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": undefined,
        "expr": Object {
          "range": Array [
            201,
            220,
          ],
          "type": "CssExpression",
          "value": "url(http://example)",
        },
        "important": false,
        "prefix": Object {
          "range": Array [
            197,
            200,
          ],
          "type": "Identifier",
          "value": "bg-",
        },
        "range": Array [
          197,
          221,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "bg-[url(http://example)]",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": Object {
          "closed": true,
          "opacity": Object {
            "range": Array [
              249,
              251,
            ],
            "type": "Identifier",
            "value": ".2",
          },
          "range": Array [
            248,
            252,
          ],
          "type": "WithOpacity",
        },
        "expr": Object {
          "range": Array [
            230,
            246,
          ],
          "type": "CssExpression",
          "value": "rgba(3 3 3 / .8)",
        },
        "important": false,
        "prefix": Object {
          "range": Array [
            224,
            229,
          ],
          "type": "Identifier",
          "value": "text-",
        },
        "range": Array [
          224,
          252,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "text-[rgba(3 3 3 / .8)]/[.2]",
      "variants": Array [],
    },
  ],
  "notClosed": Array [
    Object {
      "closed": false,
      "expressions": Array [
        Object {
          "important": false,
          "range": Array [
            265,
            271,
          ],
          "type": "ClassName",
          "value": "q-test",
        },
      ],
      "important": false,
      "range": Array [
        263,
        271,
      ],
      "type": "Group",
    },
  ],
}
`;
